### PR TITLE
BUGFIX: Enable thread-safe flags for Boost::compute upstream

### LINKED ISSUE #2016: Crash with AF OpenCL backend when multithreading
Good evening all
I experiment this crash when multi-threading my application through 2 threads doing misc linear algebra with AF :

ArrayFire v3.5.1 (OpenCL, 64-bit Linux, build 0a675e8)
[0] INTEL: Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz, 15925 MB
...
Stack dump:
0    Running pass 'Function Pass Manager' on module 'main'.
1    Running pass 'Live Interval Analysis' on function '@KER1770127750453312350'
Command terminated by signal 11


Reproduced in debug build with gdb (with my application) : the crashing thread is in af_sort() :  

1   std::__detail::_List_node_base::_M_unhook list.cc               142  0x7ffff78c92ba 
2   std::list<std::pair<std::string, std::string>>::_M_erase stl_list.h            1697 0x7fffbcead3c3 
3   std::list<std::pair<std::string, std::string>>::erase list.tcc              154  0x7fffbcead3c3 
4   boost::compute::detail::lru_cache<std::pair<std::string, std::string>, boost::compute::program>::get  lru_cache.hpp         96   0x7fffbcead3c3 
5   boost::compute::program_cache::get program_cache.hpp     96   0x7fffbcead3c3 
6   boost::compute::program_cache::get_or_build program_cache.hpp     134  0x7fffbcead3c3 
7   boost::compute::detail::meta_kernel::compile meta_kernel.hpp       347  0x7fffbceae5df 
8   boost::compute::detail::block_insertion_sort<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>> merge_sort_on_cpu.hpp 216  0x7fffbcec78eb 
9   boost::compute::detail::merge_sort_by_key_on_cpu<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>> merge_sort_on_cpu.hpp 335  0x7fffbe16877e 
10  boost::compute::detail::dispatch_sort_by_key<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>>     sort_by_key.hpp       113  0x7fffbe16877e 
11  boost::compute::sort_by_key<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>>                      sort_by_key.hpp       134  0x7fffbe169220 
12  opencl::kernel::sort0ByKeyIterative<float, unsigned int>                                                                                                                sort_by_key_impl.hpp  135  0x7fffbe169220 
13  opencl::kernel::sort0ByKey<float, unsigned int>                                                                                                                         sort_by_key_impl.hpp  254  0x7fffbe178d41 
14  opencl::sort_index<float>                                                                                                                                               sort_index.cpp        32   0x7fffbd93fc48 
15  sort_index<float>                                                                                                                                                       sort.cpp              77   0x7fffbdefe5e2 
16  af_sort_index                                                                                                                                                           sort.cpp              98   0x7fffbdefe5e2 
17  unified::AFSymbolManager::call<void * *, void * *, void *, unsigned int, bool> symbol_manager.hpp    88   0x7fffe7082c4a 
18  af_sort_index algorithm.cpp         129  0x7fffe70777e9 
19  af::sort sort.cpp              26   0x7fffe726da78 

and another thread is also in af_sort()  but trying to lock:
1  __pthread_mutex_trylock  pthread_mutex_trylock.c  35   0x7ffff7bc3fe2 
2  ?? 0x7fff7d5cfa32 
3  ?? 0x7fff7d5be0c3 
4  ?? 0x7fff7dd22d65 
5  ?? 0x7fff7dd2663e 
6  ?? 0x7fff7dd2033e 
7  ?? 0x7fff7dd009ef 
8  clCreateKernel 0x7fff7dcc9088 
9  boost::compute::kernel::kernel kernel.hpp               62   0x7fffbcea34e7 
10 boost::compute::program::create_kernel   kernel.hpp               349  0x7fffbceae5fd 
11 boost::compute::detail::meta_kernel::compile  meta_kernel.hpp          350  0x7fffbceae5fd 
12 boost::compute::detail::block_insertion_sort<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>> merge_sort_on_cpu.hpp    216  0x7fffbcec78eb 
13 boost::compute::detail::merge_sort_by_key_on_cpu<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>> merge_sort_on_cpu.hpp    335  0x7fffbe16877e 
14 boost::compute::detail::dispatch_sort_by_key<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>> sort_by_key.hpp          113  0x7fffbe16877e 
15 boost::compute::sort_by_key<boost::compute::buffer_iterator<float>, boost::compute::buffer_iterator<unsigned int>, boost::compute::greater<float>>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             sort_by_key.hpp          134  0x7fffbe169220 
16 opencl::kernel::sort0ByKeyIterative<float, unsigned int> sort_by_key_impl.hpp     135  0x7fffbe169220 
17 opencl::kernel::sort0ByKey<float, unsigned int> sort_by_key_impl.hpp     254  0x7fffbe178d41 
18 opencl::sort_index<float> sort_index.cpp           32   0x7fffbd93fc48 
19 sort_index<float> sort.cpp                 77   0x7fffbdefe5e2 
20 af_sort_index sort.cpp                 98   0x7fffbdefe5e2 
21 unified::AFSymbolManager::call<void * *, void * *, void *, unsigned int, bool>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 symbol_manager.hpp       88   0x7fffe7082c4a 
22 af_sort_index algorithm.cpp            129  0x7fffe70777e9 
23 af::sort sort.cpp                 26   0x7fffe726da78 


Kind

### PR DESCRIPTION
**Pull request by 9prady9**
Fixes #2016

### CODE REVIEW

**Comment by umar456** [comment]
Can you add a unit test which confirms this will address #2016?

File: test/threading.cpp [hunk 1]
```diff
@@ -653,3 +653,25 @@ TEST(Threading, DISABLED_MemoryManagerStressTest)
     t.join();
   }
 }
+
+TEST(Threading, BoostCompute)
```
Comment (umar456) [inline test/threading.cpp]: Call the test sort not BoostCompute

File: test/threading.cpp [hunk 2]
```diff
@@ -653,3 +653,25 @@ TEST(Threading, DISABLED_MemoryManagerStressTest)
...
+    vector<std::thread> tests;
+
+    ASSERT_EQ(AF_SUCCESS, af_set_device(0));
+
+    for (int i=0; i<THREAD_COUNT; ++i) {
+        tests.emplace_back([] {
+            for (int k=0; k<100; ++k) {
+                array a = randu(100, 100);
+                array b = sort(a);
+                array c = sort(a, 1, false);
```
Comment (umar456) [inline test/threading.cpp]: Try to test only one function. either sort or sort dim

File: test/threading.cpp [hunk 3]
```diff
@@ -653,3 +653,25 @@ TEST(Threading, DISABLED_MemoryManagerStressTest)
...
+    cleanSlate(); // Clean up everything done so far
+
+    vector<std::thread> tests;
+
+    ASSERT_EQ(AF_SUCCESS, af_set_device(0));
+
+    for (int i=0; i<THREAD_COUNT; ++i) {
+        tests.emplace_back([] {
+            for (int k=0; k<100; ++k) {
+                array a = randu(100, 100);
```
Comment (umar456) [inline test/threading.cpp]: You can probably create this out of the for loop as we are not trying to test randu with this test.

File: test/threading.cpp (hunk 2 above)
Comment (9prady9) [reply test/threading.cpp]: 👍

File: test/threading.cpp (hunk 3 above)
Comment (9prady9) [reply test/threading.cpp]: 👍

**Pull request merged**